### PR TITLE
feat(Interaction): override controller hide + use only if grabbed

### PR DESCRIPTION
**Pull request by gpvigano**
Add to Interactable Object a "Hide on touch/grab/use" option which
overrides the controller option (see issue #79 and #346).

Add to Interactable Object a "Use Only If Grabbed" option which can
prevent an object to be used if not grabbed (see issue #346).

Set "Hide Controller On Use" to "Override Hide" and checked
"Use Only If Grabbed" in the object "Usable Grabbable Sphere"
(script "Use Rotate") in the example scene
036_Controller_CustomCompoundPointer. The sphere can be used only if
grabbed and when it is used the grabbing controller disappears.

### CODE REVIEW

File: Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs [hunk 1]
```diff
@@ -182,7 +182,9 @@ protected virtual void PointerIn()
             OnDestinationMarkerEnter(SetDestinationMarkerEvent(pointerContactDistance, pointerContactTarget, destinationPosition, controllerIndex));
 
             interactableObject = pointerContactTarget.GetComponent<VRTK_InteractableObject>();
-            if (interactableObject && interactableObject.pointerActivatesUseAction && interactableObject.holdButtonToUse)
+            bool cannot_use_because_not_grabbed = (interactableObject.useOnlyIfGrabbed && !interactableObject.IsGrabbed());
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs]: pascal case your variables (not snake case)

so should be `cannotUseBecauseNotGrabbed`

but i'd also reword this to:

`useOnlyIfGrabbed` (then it's the negative of you original question so just keep that in mind :))

Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs]: also you need to define the access modifier. so this should be

`private bool <varname>`

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs [hunk 2]
```diff
@@ -24,9 +24,10 @@ public class VRTK_InteractUse : MonoBehaviour
         public event ObjectInteractEventHandler ControllerUseInteractableObject;
         public event ObjectInteractEventHandler ControllerUnuseInteractableObject;
 
-        GameObject usingObject = null;
-        VRTK_InteractTouch interactTouch;
-        VRTK_ControllerActions controllerActions;
+        private GameObject usingObject = null;
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs]: good catch! I can't believe these didn't have access modifiers on them. slap my wrists! :)

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs [hunk 3]
```diff
@@ -201,6 +203,10 @@ private void DoStartUseObject(object sender, ControllerInteractionEventArgs e)
 
             if (touchedObject != null && interactTouch.IsObjectInteractable(touchedObject))
             {
+                var vrobj = touchedObject.GetComponent<VRTK_InteractableObject>();
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs]: i'd recommend naming this something else, even if it's just `vrObj`

looks a bit weird just all lowercase

File: Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs (hunk 1 above)
Comment (gpvigano) [inline Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs]: "private" for local variables???

Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs]: shhh ignore me :P

Comment (gpvigano) [inline Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs]: useOnlyIfGrabbed is already used, anyway cannotUseBecauseNotGrabbed is just a local temporary variable used to make the code more readable (in my opinion)

Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs]: ok happy for it to stay as that then :)

Comment (gpvigano) [inline Assets/SteamVR_Unity_Toolkit/Scripts/Abstractions/VRTK_WorldPointer.cs]: OK, have I really need to change cannot_use_because_not_grabbed to cannotUseBecauseNotGrabbed in this PR? Even if I seriously promise to change it in the next days with my next PR?... (please, have mercy...)

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs (hunk 3 above)
Comment (gpvigano) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs]: OK, I surrender, it will be named "interactableObjectScript".
I'm a bit scared by this severity, anyway I got you: if many people work on the same code they must be able to clearly undestand what they read.
I already changed those names in my Working Copy.
Let me know when you finish your review, so I can try to update my branch, anyway I will push in the next days another branch, so you can start merging this PR anyway or wait for the next PR (yeah... I am still scared by Git!).

Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs]: Just keep pushing when you think you've fixed stuff and i'll keep reviewing :)

If you're getting nerves and jitters with git then just pop into the slack channel, i'll help out with commands etc :)

**Comment by gpvigano** [comment]
I just updated the commit for this PR, please check it and let me know...

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs [hunk 4]
```diff
@@ -201,6 +203,10 @@ private void DoStartUseObject(object sender, ControllerInteractionEventArgs e)
 
             if (touchedObject != null && interactTouch.IsObjectInteractable(touchedObject))
             {
+                var interactableObjectScript = touchedObject.GetComponent<VRTK_InteractableObject>();
+
+                if (interactableObjectScript.useOnlyIfGrabbed && !interactableObjectScript.IsGrabbed()) return;
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractUse.cs]: missing brackets

**Comment by thestonefox** [comment]
Mostly there now it seems, Just looks like that missing bracket issue, also the commit message needs updating so it's lowercase (the subject line should be lower case unless the word requires captialization like a name)

so should read:

`feat(Interaction): override controller hide + use only if grabbed`

Make sure to amend the commit message via git not just on github `git commit --amend`

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 9976689bfc99a94d44da676ae209b39a10f87ec5
  * Your commit message body contains a line that is longer than 72 characters
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/thestonefox/SteamVR_Unity_Toolkit

---
This message was auto-generated by https://gitcop.com

**Comment by gpvigano** [comment]
too late for amend, our network was down when you sent the last commnt and I had already committed, thus I used --force-with-lease. I hope it works.

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractableObject.cs [hunk 5]
```diff
@@ -95,6 +106,19 @@ public enum AllowedController
         private bool previousIsGrabbable;
         private bool forcedDropped;
 
+        public bool CheckHideMode(bool default_mode, ControllerHideMode override_mode )
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractableObject.cs]: All public methods require documenting in the relevant section (check the ControllerEvents section for an example of how class methods are documented)

File: Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractableObject.cs [hunk 6]
```diff
@@ -95,6 +106,19 @@ public enum AllowedController
         private bool previousIsGrabbable;
         private bool forcedDropped;
 
+        public bool CheckHideMode(bool default_mode, ControllerHideMode override_mode)
```
Comment (thestonefox) [inline Assets/SteamVR_Unity_Toolkit/Scripts/VRTK_InteractableObject.cs]: parameter names should be pascal case not snake case.

so `defaultMode` `overrideMode` and not `default_mode` `override_mode`

File: DOCUMENTATION.md [hunk 7]
```diff
@@ -1016,6 +1029,22 @@ The basis of this script is to provide a simple mechanism for identifying object
 
 ### Class Methods
 
+#### CheckHideMode/2
+
+  > `public bool CheckHideMode(bool default_mode, ControllerHideMode override_mode)`
+
+  * Parameters
+   * `bool default_mode` - The default setting of the controller (true=hide, false="don't hide").
```
Comment (thestonefox) [inline DOCUMENTATION.md]: Don't forget to update the docs with the pascal case param names too :)

Comment (gpvigano) [inline DOCUMENTATION.md]: Done. Some habits are difficult to break...

**Comment by thestonefox** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: 
it works!
